### PR TITLE
[Delivers #87947800] myusa first last

### PR DESCRIPTION
**Pull request by pkarman**
Story: https://www.pivotaltracker.com/story/show/87947800

If the User has configured MyUSA to share first and last name, the app will now use those values to populate the local User record.

Testing:

1. Logout of app
1. Logout of https://alpha.my.usa.gov/
1. Login to app
1. On MyUSA confirmation page, select **yes** for first and last names.
1. Confirm name is populated on app login

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -106,7 +106,11 @@ def self.for_email_with_slug(email, client_slug)
 
   def self.from_oauth_hash(auth_hash)
     user_data = auth_hash.extra.raw_info.to_hash
-    self.find_or_create_by(email_address: user_data['email'])
+    user = self.for_email(user_data['email'])
```
Comment (jessieay) [inline app/models/user.rb]: I'm surprised rubocop isn't picking up singe quotes?

File: spec/requests/set_user_after_login_spec.rb [hunk 2]
```diff
@@ -11,7 +11,10 @@
       get '/auth/myusa/callback'
     }.to change { User.count }.by(1)
 
-    expect(User.last.email_address).to eq('george-test@example.com')
+    new_user = User.last
+    expect(new_user.email_address).to eq('george-test@example.com')
+    expect(new_user.first_name).to eq("Georgie")
```
Comment (jessieay) [inline spec/requests/set_user_after_login_spec.rb]: should we create a fixture without first name and last name to make sure nothing errors out when they are not present?

File: config/initializers/omniauth.rb [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
 Rails.application.config.middleware.use OmniAuth::Builder do
   provider :myusa, MYUSA_KEY, MYUSA_SECRET, {
-    scope: 'profile.email',
+    scope: 'profile.email profile.first_name profile.last_name',
```
Comment (jessieay) [inline config/initializers/omniauth.rb]: That Was Easy:tm:

**Comment by jessieay** [comment]
One testing question; once that is addressed looks good to :ship:

File: config/initializers/omniauth.rb (hunk 3 above)
Comment (afeld) [inline config/initializers/omniauth.rb]: Note that this may(?) require all users to re-authorize, which is a bummer.

Comment (pkarman) [inline config/initializers/omniauth.rb]: IME, yes, it will trigger a re-authorization.

File: spec/requests/set_user_after_login_spec.rb (hunk 2 above)
Comment (pkarman) [inline spec/requests/set_user_after_login_spec.rb]: I believe the next test in that same file does that.

Comment (jessieay) [inline spec/requests/set_user_after_login_spec.rb]: I guess what I was thinking was set up something like this that does not include fname / lname and log in with that: https://github.com/18F/C2/blob/b25faa52f4c75b8d1e56127251a805ea38e055f0/spec/support/integration_spec_helper.rb

**Comment by pkarman** [comment]
@jessieay test added in 0c19ca0

**Comment by jessieay** [comment]
woo hoo! will QA now

**Comment by jessieay** [comment]
Just followed QA instructions and when I logged back in w myUSA I was not prompted to say yes to first and last names and when I re-logged in my name was not saved. Any ideas on why?

**Comment by pkarman** [comment]
hmmm.

Try deleting the cf-c2-dev application my https://alpha.my.usa.gov/ ?

Though when I try that, and try to log in to c2-dev.18f.gov, I am only prompted for allowing my email.

**Comment by afeld** [comment]
Hmm, I already had all of these selected for each environment on the MyUSA section:

screen shot 2015-11-09 at 3 02 31 pm

Paging @yozlet!

**Comment by jessieay** [comment]
OK, works on staging where I deployed (oops!). but my email address is not showing up in upper righthand of staging:

screen shot 2015-11-09 at 12 34 04 pm

is that expected?

**Comment by pkarman** [comment]
Yes that is expected (alas). `app/views/layouts/_header_nav.html.erb` uses `<%= current_user.email_address %>`

**Comment by pkarman** [comment]
You can confirm your name is populated by looking at your User record in /admin

**Comment by jessieay** [comment]
OK, I don't mind keeping as-is -- clearer for people w  multiple accts. One final question before merge: what *is* the purpose of pulling this info? are there other stories around using firstname and lastname?

**Comment by jessieay** [comment]
@pkarman yes, confirmed in admin. It is there! :tada:

**Comment by pkarman** [comment]
I believe the first/last name will appear in the Approver/Observer/Requester info on a proposal, if present.

**Pull request merged**